Documentation/contributors/design-proposals/installed-api.md

```diff
@@ -0,0 +1,282 @@
+)
+
+// Versioned API structure for external clients
+
+type Installed struct {
+    metav1.TypeMeta `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    Status InstalledStatus `json:"status,omitempty"`
+}
+
+type InstalledStatus struct {
+    // Use specific versions of tracked resources (v1alpha1)
+    Subscription v1alpha1.Subscription `json:"omitempty,subscription"`
+    ClusterServiceVersion v1alpha1.ClusterServiceVersion `json:"omitempty,clusterServiceVersion"`
+}
+
+// <definition of Installed list structure>
+```
+
+The internal type under package `github.com/operator-framework/operator-lifecycle-manager/pkg/package-server/apis/operators`:
+
+```go
+package operators
+
+import (
+    metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+    "github.com/operator-framework/operator-lifecycle-manager/pkg/api/apis/operators/v1alpha1"
+)
+
+// Internal type for apiserver use
+
+type Installed struct {
+    metav1.TypeMeta
+    metav1.ObjectMeta
+
+    Status InstalledStatus
+}
+
+type InstalledStatus struct {
+    // Lock-down specific versions here since they are part of a different API
+    Subscription v1alpha1.Subscription
+    ClusterServiceVersion v1alpha1.ClusterServiceVersion
+}
+
+// <definition of Installed list structure>
+```
+
+Conversions, defaults, open-api definitions, informers, clients, etc. will be autogenerated for these types using the [k8s code-generators](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/code-generator/README.md#code-generator). The set up should be a fairly simple extension of what currently exists for other `packageserver` types.
+
+The extension api-server will use the internal types, clients, and informers for the API group it owns (`packages.operators.coreos.com`).
+
+### Watching
+
+To support watching, `InstalledList` resources will be assigned a `ResourceVersion` field that is an aggregate of those found in its tracked `<Resource>List`s. The aggregate field will be a comma delimited string of tuples in the format `R0:RV0,R1:RV1,...,Ri:RVi,...,Rn:RVn`, where each `Ri` represents a unique, tracked, API list resource, and `RVi` represents its respective `ResourceVersion`. When a watch is requested, the extension api-server will parse the `ResourceVersion` into its components and open a watch for each API resource. When an event is received from one of these watches, the corresponding `Installed` resources will be `resynthesized` and emitted to the top-level `Installed` watch.
+
+### Deprecating Copied `CSV`s
+
+In the long-term, all logic generating copied `CSV`s will _eventually_ be removed.
+
+Changes removing copied `CSV`s from OLM:
+
+- Remove `CSV` copying during `OperatorGroup` reconciliation
+- Remove copied `CSV` queues, sync functions, and garbage collection
+- Update garbage collection for target namespace RBAC (can no longer add `OwnerReference`s to target `CSV`s)
```

Comment:
Can’t we have the ownerrefs point to the installed object? 

We might need to be careful to generate uids. I think CSV uid mangled with target namespace would work

Reply:
I just want to be cautious when it comes to adding ownerrefs to resources that aren't actually stored in etcd (e.g. what happens if the aggregated API is unavailable? I'm guessing GC runs on the children).

Reply:
We can review the GC code again, but IIRC GC controller won't GC if the apiserver is unavailable. The API needs to be up and respond correctly for it to process.

Reply:
I took a look and I think [you're right](https://github.com/kubernetes/kubernetes/blob/aaad86c2105bfbf98f26745bf8262ab8c357e8d3/pkg/controller/garbagecollector/garbagecollector.go#L340). 

I also agree it would be cleaner for the `Installed` resource to be the referent. There a couple of things I think we need to watch out for:
-  For GC to be responsive, we must [support `delete`, `list`, and `watch`](https://github.com/kubernetes/kubernetes/blob/aaad86c2105bfbf98f26745bf8262ab8c357e8d3/pkg/controller/garbagecollector/garbagecollector.go#L650) verbs for `Installed` resources
- To prevent GC from cleaning up the referring resources when the aggregated apiserver isn't fully initialized (serving discovery, but synthesized resources aren't yet populated), we should prevent readiness until synthetic resource caches have been populated at least once.
- Generate deterministic UIDs (your suggestion would work)

Let me know if this makes sense so I can get started amending the proposal.

Reply:
Yeah that sounds great - but how can we support `delete`?

Reply:
make it a no-op?